### ISSUE TITLE
GUI: wesnoth aborts with custom widget definition

### ISSUE THREAD

**Issue by white-haired-uncle** [wesnoth/wesnoth#8492]
### Game and System Information

- **Version:** Battle for Wesnoth v1.17.25+dev (https://github.com/wesnoth/wesnoth/commit/88befc29de91eeda13ce11a4620fbf642a215c55-Clean) x86_64
- **Downloaded from:** git
- **Build info:** g++ (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
- **OS:** Linux spartan 5.15.0-97-generic #107-Ubuntu SMP Wed Feb 7 13:26:48 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux 

Libraries
=========

Boost:             1.74
Lua:               5.4.6
OpenSSL/libcrypto: 3.0.0b-dev (runtime 3.0.0b-dev)
libcurl:           7.81.0 (runtime 7.81.0)
Cairo:             1.16.0 (runtime 1.16.0)
Pango:             1.50.6 (runtime 1.50.6)
SDL:               2.0.20 (runtime 2.0.20)
SDL_image:         2.0.5 (runtime 2.0.5)
SDL_mixer:         2.0.4 (runtime 2.0.4)

Features
========

Lua console completion:       yes
D-Bus notifications back end: yes

Current video settings
======================

SDL video drivers:        [x11] wayland KMSDRM dummy
Window size:              1920x1011
Game canvas size:         1920x1011
Final render target size: 1920x1011
Screen refresh rate:      60
Screen dpi:               69.87x69.98
Renderer:                 opengl (hw)
Maximum texture size:     16384x16384
VSync:                    off

### Description of the bug

I'm trying to learn about creating custom widget definitions by slightly modifying a toggle_button as defined in  .../data/gui/widget/toggle_button_tree_view_node.cfg.  And on my first attempt...

wesnoth: src/gui/widgets/styled_widget.hpp:276: gui2::canvas& gui2::styled_widget::get_canvas(unsigned int): Assertion `index < canvases_.size()' failed.
Abort(coredump)

Once again, I don't know what I'm doing here, but I managed to crash bfw.

### Steps to reproduce the behavior

1. Call provided function
2. ???
3. Profit

### Expected behavior

An error message describing the problem.

### Additional context

```
function wesnoth.wml_actions.current_test_gui()
        local definition = {
                id = "tree_view_node_custom",
                description = "Fold/unfold status indicator of a tree view node (MODIFIED).",
                wml.tag.resolution {
                        min_width = 25,
                        min_height = 19,
                        default_width = 25,
                        default_height = 19,
                        max_width = 25,
                        max_height = 19,
                        -- Selected
                        wml.tag.state_enabled {
                                wml.tag.draw {
                                        wml.tag.image { name = "buttons/fold-arrow.png~BLEND(255,0,0,1)" }
                                }
                        },
                        wml.tag.state_disabled {
                                wml.tag.draw {
                                        wml.tag.image { name = "buttons/fold-arrow.png~GS()" }
                                }
                        },
                        wml.tag.state_focused {
                                wml.tag.draw {
                                        wml.tag.image { name = "buttons/fold-arrow-pressed.png" }
                                }
                        },
                        -- Selected
                        wml.tag.state_enabled {
                                wml.tag.draw {
                                        wml.tag.image { name = "buttons/unfold-arrow.png" }
                                }
                        },
                        wml.tag.state_disabled {
                                wml.tag.draw {
                                        wml.tag.image { name = "buttons/unfold-arrow.png~GS()" }
                                }
                        },
                        wml.tag.state_focused {
                                wml.tag.draw {
                                        wml.tag.image { name = "buttons/unfold-arrow-pressed.png" }
                                }
                        },
                }
        }
        gui.add_widget_definition("toggle_button", "tree_view_node_custom", definition)

        local dialogDefinition = {
                wml.tag.tooltip { id = "tooltip_large" },
                wml.tag.helptip { id = "helptip_large" },
                wml.tag.grid {
                        wml.tag.row {
                                wml.tag.column {
                                        wml.tag.toggle_button {
                                                definition = "default"
                                        }
                                },
                                wml.tag.column {
                                        wml.tag.toggle_button {
                                                definition = "tree_view_node"
                                        }
                                },
                                wml.tag.column {
                                        wml.tag.toggle_button {
                                                definition = "tree_view_node_custom"
                                        }
                                }
                        }
                }
        }
        gui.show_dialog(dialogDefinition)
end
```

**Comment by white-haired-uncle** [comment]
I'm a bit confused by the syntax, I guess the first state_enable refers to the button when selected and the second to the button unselected?  If so, kind on non-intuitive IMO.  I'm just copying from toggle_button_tree_view_node.cfg, while merging the [state] and [enabled] tags, for example, based on an example in LotI.

I'll try to narrow down on what is causing the abort, but I figured it get it documented before I confuse myself further.

**Comment by CelticMinstrel** [comment]
> I guess the first state_enabled refers to the button when selected and the second to the button unselected?

…what on earth are you even _doing_? The toggle button does not support `[state_enabled]`, `[state_disabled]`, and `[state_focused]` tags.

> I'm just copying from toggle_button_tree_view_node.cfg, while merging the [state] and [enabled] tags, for example, based on an example in LotI.

I'd guess that the example from LotI was not a toggle button. A button, for example, _does_ use `[state_enabled]` and related tags. A toggle button however does not.

You should probably always prioritize something you see in mainline over something you see in an add-on if you're not quite sure what you're doing.

----

Basically, you're doing something that's not expected to work. _However_, it should still not crash – it should give you a readable error message so that you know what you did wrong. So, we'll leave this issue open.

**Comment by white-haired-uncle** [comment]
> > I guess the first state_enabled refers to the button when selected and the second to the button unselected?
> 
> …what on earth are you even _doing_? The toggle button does not support `[state_enabled]`, `[state_disabled]`, and `[state_focused]` tags.me when, as you said yourself, you've merged the "state" tags in with their subtags.

At first I just copied from .../data/gui/widget/toggle_button_tree_view_node.cfg so I had

```
wml.tag.state {
     wml.tag.enabled {
```
And I got some error that led me to believe that the syntax was just a little different between WML and the lua API, and when I switch to wml.tag.state_enabled the error went away.  I might have had button and not toggle_button in the gui.add_widget_definition at the time I suppose.

If the tags supported by widgets are listed somewhere, I don't know where.  I seem to remember something like this where the docs are "backwards" -- you can look for [state_enabled] and find what widgets support it, but not the other way around.  Closest thing I've found would be devdocs, which says that state_enabled is a valid state from a toggle button (which probably does not mean it's a valid TAG, but for most of us that distinction may not be completely clear).

Anyway, with this new knowledge I'll go back and see if I can get the toggle button custom def working.  Thanks.

**Comment by CelticMinstrel** [comment]
Other than the schema file, I'm not sure if there's actually a list of what tags are valid where in GUI2 WML.
